### PR TITLE
Add a META file to make ocamlfind happy

### PR DESCRIPTION
**Pull request by samoht**
Beware, I've not tested it ...

### CODE REVIEW

**Comment by ghost** [comment]
I'm not sure the META will be installed at the right location.

**Comment by samoht** [comment]
It is supposed to go in `lib/camlp4/`

**Comment by samoht** [comment]
Hum, right, this is buggy:

```
[...]
don't install META
```

File: build/install.sh [hunk 1]
```diff
@@ -139,6 +139,7 @@ for dir in Camlp4Parsers Camlp4Printers Camlp4Filters Camlp4Top; do
     $CAMLP4DIR/$dir
 done
 installdir \
+  META \
```
Comment (ghost) [inline build/install.sh]: Should be `camlp4/META`

**Comment by ghost** [comment]
The problem I'm thinking about is this one:

``` sh
$ ls `opam config var prefix`/lib/camlp4
META
$ ls `opam config var prefix`/lib/ocaml/camlp4
Camlp4Bin.cmi
Camlp4Bin.cmo
Camlp4Bin.cmx
...
# no META file
```

With this patch the META file will be installed in `opam-prefix/lib/ocaml/camlp4` instead of `opam-prefix/lib/camlp4`, so ocamlfind might not see it.

**Comment by samoht** [comment]
Haaa, you are right. Fixing this (while I'm installing ocaml.4.02 to testing this a bit more seriously)

**Comment by samoht** [comment]
Would it make sense to install camlp4 lib files in `$(opam config var prefix)/lib/camlp4` btw ?

**Comment by ghost** [comment]
I think it would break projects that don't use findlib.

**Comment by ghost** [comment]
One simple solution would be to have a separate target to install just the META file with ocamlfind.

**Comment by samoht** [comment]
Seems to be ok now.

**Comment by ghost** [comment]
OK, just one thing: since this file is copied from findlib, should it include a copyright/license/...?

**Comment by whitequark** [comment]
Should we coordinate the removal of camlp4 support from ocamlfind with Gerd? /cc @gasche

**Comment by samoht** [comment]
@whitequark well, we are not removing the support of camlp4 from ocamlfind; here we add the support of ocamlfind to camlp4, inside camlp4 itself (instead of delegating to ocamlfind). I agree we can coordinate a bit better though :-)

**Comment by whitequark** [comment]
@samoht I mean, as I understand it, currently ocamlfind will install META for camlp4 itself. The problem stems from the fact that ocamlfind will be usually installed before camlp4, not that there are no means to install camlp4's META.

File: camlp4/META [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+# or implied, including but not limited to the warranties of
+# merchantability, fitness for a particular purpose and noninfringement.
+# In no event shall Gerd Stolpmann be liable for any claim, damages or
+# other liability, whether in an action of contract, tort or otherwise,
+# arising from, out of or in connection with the Software or the use or
+# other dealings in the software.
+
+# Specifications for the "camlp4" preprocessor:
+requires = ""
+version = "4.02"
```
Comment (avsm) [inline camlp4/META]: This should source `config.sh` and substitute in `${version}`

Comment (samoht) [inline camlp4/META]: done

Comment (samoht) [inline camlp4/META]: I'm not convinced that's the good approach though, as the camlp4 version number is now supposed to be rather independent of the OCaml one.

Comment (avsm) [inline camlp4/META]: It's correct for now though, while camlp4 tracks 4.02dev.

**Comment by ghost** [comment]
I haven't tested but this might break the installation of ocamlfind if it is installed after camlp4.

**Comment by samoht** [comment]
ocamlfind will overwrite the META file with its own if it has been installed after camlp4. Appart that, there is no breakage.

**Comment by gerdstolpmann** [comment]
Merged this to the official svn repo.

My thinking is that there is no perfect way to do this. For packagers this is already a good solution. For other users, it is only important that they are guided the right way, and if you install in the order 1. ocaml, 2. findlib, 3. camlp4, everything will be fine by running "configure" without options. If you install 1. ocaml, 2. camlp4, 3. findlib, there will be hopefully a warning by camlp4 that there isn't findlib support (haven't checked, though).

**Pull request merged**